### PR TITLE
cilium-operator: support subnets filters

### PR DESCRIPTION
**Pull request by bpineau**
The `cilium-operator` start/resync time and memory consumption are excessive in our largest AWS accounts, due to the `GetInstances` regularly fetching each VPC's interfaces.
We run our clusters in dedicated subnets, so the set of relevant interfaces can be easily reduced.

In one of our AWS accounts, fetching all subnets and interfaces requires about 2GB of memory, and takes about 60s.
Filtering on cluster's subnets, we're down to 20MB and 2s for a 130 nodes cluster (or 200MB / 8s for 2300 nodes).

### CODE REVIEW

File: pkg/aws/ec2/ec2.go [hunk 1]
```diff
@@ -43,14 +44,35 @@ type MetricsAPI interface {
...
-		limiter:    helpers.NewApiLimiter(metrics, rateLimit, burst),
+		ec2Client:      ec2Client,
+		metricsAPI:     metrics,
+		limiter:        helpers.NewApiLimiter(metrics, rateLimit, burst),
+		subnetsFilters: subnetsFilters,
 	}
 }
 
+func NewSubnetsFilters(tags map[string]string, ids []string) []ec2.Filter {
+	var filters []ec2.Filter
```
Comment (christarazi) [inline pkg/aws/ec2/ec2.go]: We can preallocate the `filters` slice here because we know the size of `tags` at call-time. This will help our goal of reducing memory usage as seen here: #10716 :smile:

**Comment by christarazi** [comment]
test-me-please

File: pkg/option/config.go [hunk 2]
```diff
@@ -1570,6 +1576,12 @@ type DaemonConfig struct {
 	// ENITags are the tags that will be added to every ENI created by the AWS ENI IPAM
 	ENITags map[string]string
 
+	// SubnetsIDs are optional subnets IDs used to filter subnets and interfaces listing
+	SubnetsIDs []string
```
Comment (tgraf) [inline pkg/option/config.go]: Please rename to something that indicates that this is being used for IPAM. Same for `SubnetTags`.

File: pkg/aws/ec2/ec2.go [hunk 3]
```diff
@@ -80,6 +103,18 @@ func (c *Client) describeNetworkInterfaces(ctx context.Context) ([]ec2.NetworkIn
 			req.NextToken = &nextToken
 		}
 
+		if len(c.subnetsFilters) > 0 {
+			subnetsIDs := make([]string, len(subnets))
```
Comment (tgraf) [inline pkg/aws/ec2/ec2.go]: You probably want `make([]string, 0, len(subnets))`

File: pkg/aws/ec2/ec2.go [hunk 4]
```diff
@@ -43,14 +44,35 @@ type MetricsAPI interface {
...
-		metricsAPI: metrics,
-		limiter:    helpers.NewApiLimiter(metrics, rateLimit, burst),
+		ec2Client:      ec2Client,
+		metricsAPI:     metrics,
+		limiter:        helpers.NewApiLimiter(metrics, rateLimit, burst),
+		subnetsFilters: subnetsFilters,
 	}
 }
 
+func NewSubnetsFilters(tags map[string]string, ids []string) []ec2.Filter {
```
Comment (tgraf) [inline pkg/aws/ec2/ec2.go:58]: Please add a comment and a unit test

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks great! A couple of comments inline. I think it makes sense to assume that we'll want this for Azure and GCP mode as well later on so the naming should be generic enough so we won't have to rename later on.

**Review by christarazi — CHANGES_REQUESTED** [review]

File: pkg/aws/ec2/ec2_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := NewSubnetsFilters(tt.args.tags, tt.args.ids); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("NewSubnetsFilters() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+	_ = tests
```
Comment (christarazi) [inline pkg/aws/ec2/ec2_test.go]: Why is this needed?

**Review by christarazi — CHANGES_REQUESTED** [review]
Some minor changes, but otherwise LGTM!

File: Documentation/cmdref/cilium-operator.md [hunk 6]
```diff
@@ -49,6 +49,8 @@ cilium-operator [flags]
       --operator-api-serve-addr string          Address to serve API requests (default "localhost:9234")
       --operator-prometheus-serve-addr string   Address to serve Prometheus metrics (default ":6942")
       --parallel-alloc-workers int              Maximum number of parallel IPAM workers (default 50)
+      --subnets-ids strings                     Subnets IDs (separated by commas)
+      --subnets-tags stringToString             Subnets tags in the form of k1=v1,k2=v2 (multiple k/v pairs can also be passed by repeating the CLI flag (default [])
```
Comment (ungureanuvladvictor) [inline Documentation/cmdref/cilium-operator.md]: What about naming this `subnet-tags-filter` ? We already have `eni-tags` which is used to be passed when creating AWS ENIs and don't want to have some confusions that the operator will create subnets as well.

**Comment by ungureanuvladvictor** [comment]
@bpineau all good from my side (just wanted the cli arg clarification). Will let @tgraf + @christarazi finish the review here, don't block on me.

File: pkg/aws/ec2/ec2.go [hunk 7]
```diff
@@ -43,14 +44,42 @@ type MetricsAPI interface {
...
 
+// NewSubnetsFilters transforms a map of tags and values and a slice of subnets
+// into a slice of ec2.Filter adequate to filter AWS subnets.
+func NewSubnetsFilters(tags map[string]string, ids []string) []ec2.Filter {
+	var filters []ec2.Filter
+	if len(ids) == 0 {
+		filters = make([]ec2.Filter, 0, len(tags))
+	} else {
+		filters = make([]ec2.Filter, 0, len(tags)+1)
+	}
```
Comment (christarazi) [inline pkg/aws/ec2/ec2.go]: We can just take `filters = make([]ec2.Filter, 0, len(tags)+1)`. There's no need to have a branch to optimize the preallocation by one element.

**Review by christarazi — APPROVED** [review]

File: pkg/aws/ec2/ec2.go (hunk 4 above)
Comment (bpineau) [reply pkg/aws/ec2/ec2.go:58]: Added both (though github UI don't show the comment as "outdated")

**Comment by bpineau** [comment]
I think the failing integration tests are unrelated/flacky.
Would a 1.7 backport be conceivable (if/when this hits master indeed)?

**Comment by aanm** [comment]
@bpineau please rebase this PR with master as some flakes are fixed in master.

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Docs changes LGTM.

**Review by tgraf — APPROVED** [review]

**Comment by bpineau** [comment]
Rebased and tested over the current master HEAD (@ungureanuvladvictor you're listed as pending reviewer, mind having a loook?).

**Review by ungureanuvladvictor — APPROVED** [review]
This all looks great to me! Sorry for being late to this :(

**Comment by ungureanuvladvictor** [comment]
test-me-please

**Comment by joestringer** [comment]
Travis failed with something that looks relevant:

```
--- FAIL: TestNewSubnetsFilters (0.00s)
    --- PASS: TestNewSubnetsFilters/empty_arguments (0.00s)
    --- PASS: TestNewSubnetsFilters/ids_only (0.00s)
    --- FAIL: TestNewSubnetsFilters/tags_only (0.00s)
    --- PASS: TestNewSubnetsFilters/tags_and_ids (0.00s)
FAIL
coverage: 3.1% of statements in ./...
```

Is there a potential for race conditions in the test?

**Comment by bpineau** [comment]
@joestringer Yes, I fixed it to be stable (hence the sorts in the test) after the previous run.

**Comment by joestringer** [comment]
Hmm, did you already push those changes? Travis triggers automatically when you push your code, so the current failure reported by Travis should be triggered against the current version of the PR. That suggests that either the PR as-is still has another flake in it, or the code here isn't the latest version. Unless I'm missing some other possibility :)

**Comment by bpineau** [comment]
You guessed correctly, I forgot to push (just did now) 🤦‍♂

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit known flake #10821 in `Cilium-Tests-With-Kernel`, safe to ignore
